README.md
# consul-rs-plus


A plus consul client package for Rust, more functions for microservice.

- [x] key/value operation
- [x] session operation
- [x] key/value & session combination operation
- [x] key/value folder values(for micro service nodes info)
- [x] watch keys tree change function(usual use in microservice)


## install
set in Cargo dependencies
```toml
[dependencies]
consul-rs-plus = "0.1.7"
```

## Usage
```rust
extern crate consul_rs_plus;
use consul_rs_plus::Client;

fn main() {
    let mut c = Client::new("localhost", 8500);
    // debug enable
    c.debug = true;

    let ok = c.kv_set("test-key", "test_value").unwrap();
    assert_eq!(ok, true);

    let kvpairs = c.kv_get("test-key").unwrap();
    let kvpair = &kvpairs[0];
    let v = kvpair.get_value().unwrap();
    assert_eq!(b"test_value"[..].to_vec(), v);
  
    let ok = c.kv_delete("test-key").unwrap();
    assert_eq!(ok, true);
}
```

## Test
the test case write in function code file or tests folder, all nromal operation test in lib.rs.
```rust
#[cfg(test)]
mod tests {
  use crate::Client;
  use base64::Config;
  use crate::config;

  #[test]
  fn test_kv_get() {
    let host = config::CONFIG["consul_addr"];
    let client = Client::new(host, 8500);
    let my_keys = client.kv_get("my-key").unwrap();
    for k in my_keys {
      println!("k: {:?}", k);
    }
  }
  
}
```

## Micro service nodes watch
if you write the micro service framework, the service register like [rpcx-plus](https://github.com/halokid/rpcx-plus), folder tree style, for example， Echo service in folder(key, value) is 
`` /mytest/Echo/tcp@8.8.8.8:999 ``, you can find the sample code for watch service change below,
they are in code repo. you can cache the service info when the service change(the folder tree change).
```rust

#[tokio::test]
  async fn test_watch_folder_tree_tmpsc() {
  env_logger::init();
  let folder = "mytest".to_string();
  let mut nodes_service: Vec<String> = Vec::new();     // service cache
  let (sx, mut rx) = tmpsc::channel(1);
  let kv = KVPair::new();
  let client = Client::new("consul_test", 8500);
  let mut index = kv.get_folder_index(&client, &folder);
  log::info!("index orgin ------- {}", index);
  tokio::task::spawn(async move {
    loop {
      thread::sleep(time::Duration::from_secs(5));
      let mut index_ck = kv.get_folder_index(&client, &folder);
      log::info!("index_ck ------- {}", index_ck);
      if !index_ck.eq(index.as_str()) {
        log::info!("=== get newest nodes service, send coroutine ===");
        let nodes_v = kv.get_folder_allkeys(&client, &folder);
        let nodes_v_cl = nodes_v.clone();
        log::info!("[send] === in spawn nodes_v_cl: {:?}", nodes_v_cl);
        sx.send(nodes_v_cl).await.unwrap();    // todo: just make the channel full!
        index = index_ck;
      } else {
        log::info!("=== nodes_service no change ===");
      }
    }
  });
}

```
you can run this test case in code repo.





CHANGELOG.md

# v0.2.0

### Features
- se `0.11` version reqwest, all request HTTP change to use async/await for good performance
- upgrade `kv` function use async/await
- upgrade `session` function use async/await
- add Consul `service` functions implement

# v0.1.7

Initial changelog release



tests/kv_test.rs
extern crate consul_rs_plus;

use consul_rs_plus::Client;

#[test]
fn your_test() {
}


src/lib.rs
#![allow(dead_code)]

#[macro_use]
extern crate serde_derive;
extern crate base64;
extern crate reqwest;
extern crate serde_json;

pub mod kv;
mod config;
pub mod session;
pub mod pkg;
pub mod service;
pub mod vo;

use self::kv::*;
use std::io::Read;
use crate::pkg::CustomError;
use base64::{decode};
use env_logger::Env;

// todo: global use varible here
// pub const PKGX: pkg::Pkg = pkg::Pkg::new(true);

// env_logger::Builder::from_env(Env::default().default_filter_or("info")).init();
// env_logger::init();

pub struct Client {
  pub debug: bool,

  host: String,
  port: u16,

  kv:   KVPair,
  session: session::Session,
}

impl Client {
  pub fn new<S: Into<String>>(host: S, port: u16) -> Client {
    // env_logger::init();
    env_logger::Builder::from_env(Env::default().default_filter_or("info")).init();

    Client {
      debug: false,
      host: host.into(),
      port: port,
      kv: KVPair::new(),
      session: session::Session::new(),
    }
  }

  pub fn debug_print(&self, s: &str) {
    // if self.debug {
    //   println!("[DEBUG] --- {:?}", s);
    // }
    println!("[DEBUG] --- {:?}", s);
  }

  pub fn kv_get<S: Into<String>>(&self, key: S) -> String {
    let keyx = key.into();
    let keyxx = keyx.clone();
    let res = self.kv.get(self, keyx);
    match res {
      Ok(kvs) => {
        let kv = kvs.get(0).unwrap();
        let val = &kv.Value;
        let val_de = decode(val).unwrap();
        let val_de_str = String::from_utf8(val_de).unwrap();
        val_de_str
      }
      Err(err) => {
        log::error!("key {} not exists, err: {}", keyxx, err);
        // err
        "keyNoExists_or_valIsNull".to_string()
      }
    }
  }

  pub fn kv_get_folder<S: Into<String>>(&self, folder: S) -> Result<Vec<String>,
    String> {
    let folder_keys = self.kv.get_folder_keys(self, folder).unwrap();
    serde_json::from_str::<Vec<String>>(&folder_keys).map_err(|e| e.to_string())
  }

  pub fn kv_set<S: Into<String>>(&self, key: S, v: S) -> Result<bool, String> {
    self.kv.set(self, key, v)
  }

  pub fn kv_set_with_session<S: Into<String>>(&self, key: S, v: S, session: S)
                          -> Result<bool, String> {
    self.kv.set_with_session(self, key, v, session)
  }

  pub fn kv_delete<S: Into<String>>(&self, key: S) -> Result<bool, String> {
    self.kv.delete(self, key)
  }

  pub fn kv_delete_both_session<S: Into<String>>(&self, key: S) -> Result<bool, String> {
    self.kv.delete_both_session(self, key)
  }

  /*
  pub async fn kv_folder_watch<S: Into<String>>(&self, folder: S) -> Result<bool, String> {
    let res = self.kv.watch_tree(self, folder).await;

    Ok(true)
  }
   */

  pub fn session_set(&self, lock_delay: String, name: String, node: String,
                     behavior: String, ttl: String) -> String {
    let mut s = session::Session::new();
    s.LockDelay = "0.001s".to_string();
    if name != "".to_string() {
      s.name = name;
    }

    if node != "".to_string() {
      let nodex = Some(node);
      s.node = nodex;
    }
    s.Behavior = behavior;
    s.TTL = ttl;
    self.debug_print(format!("lib session set: {:?}", s).as_str(), );
    self.session.set(self, &s)
  }

  pub fn session_renew(&self, sid: &str) -> Result<(), CustomError> {
    self.session.renew(self, sid)
  }

  pub fn session_delete(&self, sid: &str) -> String {
    self.session.delete(self, sid)
  }

  pub fn service_get(&self, name: String) -> Vec<String> {
    vec![]
  }

}

#[cfg(test)]
mod tests {
  use crate::Client;
  use base64::Config;
  use tokio::sync::mpsc;
  use crate::config;
  use crate::pkg::CustomError;
  use crate::kv::KVPair;

  /*
  // #[test]
  #[tokio::test]
  async fn test_kv_folder_watch() {
    env_logger::init();
    for i in 0..3 {
      log::info!("i --- {:?}", i);
    }
    log::info!("log输出");
    let host = config::CONFIG["consul_addr"];
    let client = Client::new(host, 8500);
    let res = client.kv_folder_watch("foo").await;
  }
   */

  /*
  #[tokio::test]
  async fn test_kv_folder_watch_getnodes() {
    env_logger::init();
    for i in 0..3 {
      log::info!("i --- {:?}", i);
    }
    log::info!("log输出");
    let host = config::CONFIG["consul_addr"];
    let client = Client::new(host, 8500);
    let mut res = false;
    let (sx, mut rx) = mpsc::channel(1);

    tokio::task::spawn({
      let resx = client.kv_folder_watch("foo").await.unwrap();
      sx.send(resx).await.unwrap();
    });

    while let Some(resx) = rx.recv().await {
      log::info!("resx --- {:?}", resx);
    }
  }
   */

  #[test]
  fn test_kv_delete_both_session() {
    let host = config::CONFIG["consul_addr"];
    let client = Client::new(host, 8500);
    let res = client.kv_delete_both_session("my-key");
    println!("res ---------- {}", res.unwrap());
  }

  #[test]
  fn test_kv_get() {
    // let host = config::CONFIG["consul_addr"];
    let host = "consul_test";
    let client = Client::new(host, 8500);
    let val = client.kv_get("my-key");
    if val.eq("keyNoExists_or_valIsNull") {
      println!("key {} keyNoExists_or_valIsNull", "my-key");
    } else {
      println!("key {} exists", "my-key");
      println!("val ----- {}", val);
    }
  }

  #[test]
  fn test_session_set() {
    // let host = config::CONFIG["consul_addr"];
    let host = "consul_test";
    let client = Client::new(host, 8500);
    let se = client.session_set("15s".to_string(), "my-session".to_string(), "node1".to_string(), "release".to_string(), "10m0s".to_string());
  }

  #[test]
  fn test_kv_set_with_session() {
    let host = config::CONFIG["consul_addr"];
    let client = Client::new(host, 8500);
    let session = client.session_set("15s".to_string(), "my-session".to_string(), "node1".to_string(), "release".to_string(), "10m0s".to_string());

    let res = client.kv_set_with_session("my-key", "my-val", session.as_str()).unwrap();
    println!("res ------- {}", res);
  }

  #[test]
  fn test_session_renew() {
    let host = config::CONFIG["consul_addr"];
    let client = Client::new(host, 8500);
    let ok = client.session_renew("d5663534-82f9-429b-954c-ae63d59d3502");
    match ok {
      Ok(_) => { println!("---ok---");}
      Err(_) => { println!("---err---");}
    }
  }

  #[test]
  fn test_kv_ge_folder() {
    let host = config::CONFIG["consul_addr"];
    let client = Client::new(host, 8500);
    let res = client.kv_get_folder("pomid/Echo");
    let svcs = res.unwrap();
    println!("svcs ------------- {:?}", svcs);
  }

}






src/config.rs

use lazy_static::*;
use std::collections::HashMap;
use std::env;
// use env_logger::Env;

lazy_static! {
  // env_logger::Builder::from_env(Env::default().default_filter_or("info")).init();

  pub static ref CONFIG: HashMap<&'static str, &'static str> = {
    let mut config = HashMap::new();

    let cmd_args: Vec<_> = env::args().collect();
    let mut run_env = "";
    if cmd_args.len() > 1 {
      run_env = &cmd_args[1];
    } else {
      run_env = "test"
    }

    run_env = "test";   // just for test
    match run_env {
      "dev" => {
        config.insert("consul_addr", "127.0.0.1");
        config.insert("consul_port", "8500");
      },
      "test" => {
        config.insert("consul_addr", "127.0.0.1");
        config.insert("consul_port", "8500");
      },
      "prd" => {
        config.insert("consul_addr", "8.8.8.8");
        config.insert("consul_port", "8500");
      },
      _ => {},
    }

    println!("runtime config ----- {}: {:?}", run_env, config);
    config
  };
}


src/service.rs
use std::collections::HashMap;
use std::error::Error;
use log::log;
use serde_json::Value;
use crate::Client;
use crate::pkg::CustomError;

pub struct Service {
  consul_client:  Client,
  name: String,
}

impl Service {
  pub fn new<S: Into<String>>(c: Client, name: S) -> Self {
    Service {
      consul_client: c,
      name: name.into(),
    }
  }

  /// Get the health services, base on below API url
  /// http://localhost:8500/v1/catalog/service/neon_broker
  /// http://localhost:8500/v1/health/checks/neon_broker
  pub fn get(&self, c: &Client, service_name: &str) -> Result<Vec<String>, CustomError> {
    let url = format!("http://{}:{}/v1/health/checks/{}", c.host, c.port, service_name);

    // let nodes = self._get_nodes()

    Ok(vec![])
  }

  /// return []String service_id
  pub async fn _get_nodes(&self) -> HashMap<String, String> {
    let url = format!("http://{}:{}/v1/catalog/service/{}", self.consul_client.host, self.consul_client.port, self.name);
    // println!("Fetching {:?}...", url);
    log::info!("Fetching {:?}...", url);

    // reqwest::get() is a convenience function.
    // In most cases, you should create/build a reqwest::Client and reuse
    // it for all requests.
    let rsp = reqwest::get(url).await;
    let res = rsp.unwrap();
    // println!("Response: {:?} {}", res.version(), res.status());
    // println!("Headers: {:#?}\n", res.headers());
    let body = res.text().await;
    // println!("{:?}", body);
    // body.unwrap()
    let mut nodes = HashMap::new();
    let body_js: Value = serde_json::from_str(body.unwrap().as_str()).unwrap();
    log::info!("body_js -->>> {:?}", body_js);
    // let body_js_arr = body_js.as_array();
    match body_js.as_array() {
      None => { }

      Some(_) => {
        for service in body_js.as_array().unwrap() {
          log::info!("ServiceID -->>> {:?}, {}, {}", service["ServiceID"], service["ServiceID"],
          service["ServiceID"].as_str().unwrap().to_string());
          let node_addr = format!("{}:{}",
                                  service["ServiceAddress"].as_str().unwrap().to_string(),
                                  service["ServicePort"].as_u64().unwrap().to_string());
          log::info!("node_addr -->>> {}", node_addr);
          nodes.insert(service["ServiceID"].as_str().unwrap().to_string(), node_addr);
        }
      }
    }
    nodes
  }

  /// return `service_id: staus` hashmap
  pub async fn _get_health(&self) -> Vec<String> {
    let url = format!("http://{}:{}/v1/health/checks/{}", self.consul_client.host, self.consul_client.port, self.name);
    log::info!("Fetching {:?}...", url);
    let rsp = reqwest::get(url).await;
    let res = rsp.unwrap();
    let body = res.text().await;
    // let mut node_health = HashMap::new();
    let mut nodes_health: Vec<String> = Vec::new();
    let body_js: Value = serde_json::from_str(body.unwrap().as_str()).unwrap();
    log::info!("body_js -->>> {:?}", body_js);
    match body_js.as_array() {
      None => { }

      Some(body_js_arr) => {
        for service in body_js_arr {
          let status = service["Status"].as_str().unwrap();
          log::info!("status -->>> {:?}", status);
          if status == "passing" {
            nodes_health.push(service["ServiceID"].as_str().unwrap().to_string());
          }
        }
      }

    }
    nodes_health
  }


}

#[cfg(test)]
mod tests {
  use crate::Client;
  use crate::service::Service;

  // #[test]
  // fn test_get_nodes() {
  //   let client = Client::new("consul_test", 8500);
  //   let s = Service::new();
  //   s._get_nodes(&client, "neon_broker");
  // }
}














src/session.rs
use crate::{Client};
use std::io::Read;
use std::fmt::Error;
use std::time::Duration;
use crate::pkg::CustomError;

const TIMEOUT: u64 = 600;

#[derive(Serialize, Deserialize, Debug)]
pub struct Session {
  pub id:       String,
  pub name:     String,
  pub node:     Option<String>,
  pub LockDelay: String,
  pub Behavior:  String,
  pub TTL:      String,
  pub node_checks:  Vec<String>,
  service_checks: Option<String>,
  create_index:   u32,
  modify_index:   u32,
  // timeout:        u64,
}

#[derive(Debug, Serialize, Deserialize)]
struct SessionSet {
  ID:   String,
}

impl Session {

  pub fn new() -> Self {
    Session {
      id: "".to_string(),
      name: "".to_string(),
      node: None,
      LockDelay: "".to_string(),
      Behavior: "".to_string(),
      TTL: "".to_string(),
      node_checks: vec![],
      service_checks: None,
      create_index: 0,
      modify_index: 0,
      // timeout:      600,
    }
  }

  pub fn set(&self, c: &Client, s: &Session) -> String {
    "".to_string()
  }

  pub fn renew(&self, c: &Client, sid: &str) -> Result<(), CustomError> {
    Err(CustomError("".to_string()))
  }

  pub fn delete(&self, c: &Client, sid: &str) -> String {
    "".to_string()
  }
}






src/kv.rs
use crate::{Client};
use std::io::Read;
use std::{thread, time};
use std::borrow::Borrow;
use std::collections::HashMap;
use std::prelude::v1::Option::Some;
use crate::pkg::CustomError;
use tokio::sync::mpsc;
// use std::sync::mpsc;
use std::thread::sleep;
use tokio::time::Duration;
use std::str;

const TIMEOUT: u64 = 600;

#[derive(Serialize, Deserialize)]
#[derive(Debug, Eq, PartialEq)]
#[serde(rename_all = "PascalCase")]
pub struct KVPair {
  pub Key: String,
  pub CreateIndex: u64,
  pub ModifyIndex: u64,
  pub LockIndex: u64,
  pub Flags: u64,
  pub Value: String,
  #[serde(default = "default_string")]
  pub Session: String,
  // timeout:        u64,
}

fn default_string() -> String {
  "".to_string()
}

impl KVPair {
  pub fn new() -> Self {
    KVPair {
      Key: "".to_string(),
      CreateIndex: 0,
      ModifyIndex: 0,
      LockIndex: 0,
      Flags: 0,
      Value: "".to_string(),
      Session: "".to_string(),
      // timeout:      600,
    }
  }

  // todo: Into<String> get the ownship for varible?
  pub fn get<S: Into<String>>(&self, c: &Client, key: S) -> Result<Vec<KVPair>, String> {
    Err("".to_string())
  }

  pub fn get_folder_keys<S: Into<String>>(&self, c: &Client, key: S) -> Result<String,
    String> {
    Err("".to_string())
  }

  pub fn set<S: Into<String>>(&self, c: &Client, key: S, v: S) -> Result<bool, String> {
    Err("".to_string())
  }

  pub fn set_with_session<S: Into<String>>(&self, c: &Client, key: S, v: S, session: S)
                                           -> Result<bool, String> {
    Err("".to_string())
  }

  pub fn delete<S: Into<String>>(&self, c: &Client, key: S) -> Result<bool, String> {
    Err("".to_string())
  }

  pub fn delete_both_session<S: Into<String>>(&self, c: &Client, key: S) -> Result<bool, String> {
    Err("".to_string())
  }

  pub fn get_value(&self) -> Result<Vec<u8>, base64::DecodeError> {
    base64::decode(&self.Value)
  }

  /*
  pub async fn watch_tree<S: Into<String>>(&self, c: &Client, folder: S,
              mut svc_nodes: HashMap<String, Vec<String>>) -> Result<bool, String> {
    log::info!("watch_tree log输出 1");
    println!("watch_tree log输出 2");
    log::info!("watch_tree log输出 3");
    let hostx = c.host.clone();
    let portx = c.port.clone();
    let url = format!("http://{}:{}/v1/kv/{}", c.host, c.port, folder.into());
    let mut rsp = reqwest::get(&url).map_err(|e| e.to_string())?;
    let header = rsp.headers();
    log::info!("header ----- {:?}, {:?}", header, header.get("x-consul-index").unwrap());
    let origin_index = header.get("x-consul-index").unwrap();
    let (index_check, mut rx) = mpsc::channel(1);

    tokio::task::spawn(async move {
      for i in 0..3 {
        sleep(Duration::from_secs(5));
        let url = format!("http://{}:{}/v1/kv/foo", hostx, portx);
        let mut rspx = reqwest::get(&url).map_err(|e| e.to_string()).unwrap();
        let header = rspx.headers();
        log::info!("header in spawn ----- {:?}, {:?}", header, header.get("x-consul-index").unwrap());
        let check_index = header.get("x-consul-index").unwrap();
        // let check_index = "xx";
        let check_index_owen = check_index.to_owned();
        index_check.send(check_index_owen).await.unwrap();
        log::info!("=== watch tree spawn ===");
      }
    });

    while let Some(check_index) = rx.recv().await {
      log::info!("check --- {:?}", check_index);
      svc_nodes.insert("key".into(), "val".into());
    }

    sleep(Duration::from_secs(30));
    Ok(true)
  }
   */

  fn get_folder_index<S: Into<String>>(&self, c: &Client, folder: S) -> String {
    "".to_string()
  }

  // if you use cakeRabbit micro-service frmework, it use folder for service nodes
  // you can get all the service nodes use this fn, the consul API: /v1/kv/folder?keys
  fn get_folder_allkeys<S: Into<String>>(&self, c: &Client, folder: S) -> Vec<String> {
    vec![]
  }
}


#[cfg(test)]
mod tests {
  use crate::Client;
  use crate::kv::KVPair;
  use tokio::sync::mpsc as tmpsc;
  use tokio::sync::oneshot as toneshot;
  use std::sync::mpsc as smpsc;
  use std::{thread, time};
  use log;
  use chrono::prelude::*;

  #[test]
  fn test_get_folder_allkeys() {
    let kv = KVPair::new();
    let client = Client::new("consul_test", 8500);
    let nodes_v = kv.get_folder_allkeys(&client, "mytest".to_string());
    println!("node_v -------- {:?}", nodes_v);
    println!("node_v 0 -------- {:?}", &nodes_v[0..1]);
  }

  #[test]
  fn test_get_folder_index() {
    let kv = KVPair::new();
    let client = Client::new("consul_test", 8500);
    let index = kv.get_folder_index(&client, "mytest".to_string());
    println!("index ------ {}", index);
  }

  // todo: for micro service, service client should watch the service tree, load the services
  // todo: cache, dont need to request the register center every call time.
  // todo: this fn should be running in main thread. while get sx.send, then reload services cache.
  #[tokio::test]
  async fn test_watch_folder_tree_tmpsc() {
    env_logger::init();
    let folder = "mytest".to_string();
    let mut nodes_service: Vec<String> = Vec::new();     // service cache
    // todo: if the index change, send the new nodes services between coroutine
    let (sx, mut rx) = tmpsc::channel(1);
    // let (mut sx, mut rx) = toneshot::channel();
    // let (sx, mut rx) = smpsc::channel();
    let kv = KVPair::new();
    let client = Client::new("consul_test", 8500);
    let mut index = kv.get_folder_index(&client, &folder);
    log::info!("index orgin ------- {}", index);

    tokio::task::spawn(async move {
      // loop {
      for i in 0..10 {
        thread::sleep(time::Duration::from_secs(5));
        // /*
        let mut index_ck = kv.get_folder_index(&client, &folder);
        log::info!("index_ck ------- {}", index_ck);
        if !index_ck.eq(index.as_str()) {
          log::info!("=== get newest nodes service, send coroutine ===");
          let nodes_v = kv.get_folder_allkeys(&client, &folder);
          let nodes_v_cl = nodes_v.clone();
          log::info!("[send] === in spawn nodes_v_cl: {:?}", nodes_v_cl);
          sx.send(nodes_v_cl).await.unwrap();    // todo: just make the channel full!
          // sx.send(nodes_v_cl).await.unwrap();
          index = index_ck;
        } else {
          log::info!("=== nodes_service no change ===");
          // sx.send(vec![]).await.unwrap();
        }
        // */

        // log::info!("=== get newest nodes service, send coroutine ===");
        // let nodes_v = kv.get_folder_allkeys(&client, &folder);
        // let nodes_v_cl = nodes_v.clone();
        // log::info!("[send] === in spawn nodes_v_cl: {:?}", nodes_v_cl);
        // sx.send(nodes_v_cl).await.unwrap();    // todo: just make the channel full!

        // &sx.send(nodes_v_cl).unwrap();
      }
    });

    // /*
    while let Some(nodes_v) = rx.recv().await {
      log::info!("=== [一次recv开始] ===");
      log::info!("[recv 1] === recv nodes_v --- {:?}", nodes_v);
      nodes_service = nodes_v;
      log::info!("[recv 2] === reload nodes_service --- {:?}", nodes_service);
      log::info!("=== [一次recv结束] ===");
      // return nodes_v;
    }
    // */

    // todo: if channle is oneshot
    // match rx.await {
    //   Ok(v) => { println!("v --- {:?}", v); }
    //   Err(e) => { println!("err ---- {}", e); }
    // }

    // todo: will not run!!!
    log::info!("nodes_service now is --- {:?}", nodes_service)
    // return vec![];
  }

  #[tokio::test]
  async fn test_tokio_channel() {
    println!("====================================================================");
    // loop send ----------------------------------------
    let (sx, mut rx) = tmpsc::channel(1);

    tokio::task::spawn(async move {
      for i in 0..9 {
        thread::sleep(std::time::Duration::from_secs(5));
        println!("=== send by tokio spawn --- {}, {}", i, Local::now());
        sx.send(i).await.unwrap();
      }
    });

    while let Some(i) = rx.recv().await {
      // println!("5 now time is {:?}", Local::now());
      println!("=== recv from tokio spawn --- {}, {}", i, Local::now());
    }

    // sleep(Duration::from_secs(10));
    // println!("6 now time is {:?}", Local::now());
  }

  #[tokio::test]
  async fn test_watch_folder_tree_smpsc() {
    env_logger::init();
    let mut nodes_service: Vec<String> = Vec::new();     // service cache
    // todo: if the index change, send the new nodes services between coroutine
    let (sx, mut rx) = smpsc::channel();
    let kv = KVPair::new();
    let client = Client::new("consul_test", 8500);
    let mut index = kv.get_folder_index(&client, "mytest".to_string());
    log::info!("index orgin ------- {}", index);

    tokio::task::spawn(async move {
      for i in 0..90 {
        thread::sleep(time::Duration::from_secs(5));
        let mut index_ck = kv.get_folder_index(&client, "mytest".to_string());
        log::info!("index_ck ------- {}", index_ck);
        if !index_ck.eq(index.as_str()) {
          log::info!("=== get newest nodes service, send coroutine ===");
          let nodes_v = kv.get_folder_allkeys(&client, "mytest".to_string());
          sx.send(nodes_v).unwrap();
          index = index_ck;
        } else {
          log::info!("=== nodes_service no change ===");
          // sx.send(vec![]).await.unwrap();
        }
      }
    }).await;   // todo: this is await will loop run the tokio::spawn, so cannot run next process

    // todo: will not run forever!!!!!
    let nodes_v = rx.recv().unwrap();
    nodes_service = nodes_v;
    log::info!("reload nodes_service --- {:?}", nodes_service);
    // return nodes_v;

    // return vec![];
  }

  /*
  #[tokio::test]
  async fn test_get_newest_folder_tree() {
    env_logger::init();
    let nodes_v = test_watch_folder_tree().await;
    log::info!("nodes_v ---------- {:?}", nodes_v);
  }
   */


  #[test]
  fn unmarshal_kv_pair() {
    let dat = r#"[
        {
            "LockIndex":666,
            "Key":"key/path",
            "Flags":0,
            "Value":"dHJ1ZQ==",
            "CreateIndex":495627,
            "ModifyIndex":495627
        }]"#;
    let v: Vec<super::KVPair> = serde_json::from_str(dat).unwrap();
    assert_eq!(v[0].LockIndex, 666);
    assert_eq!(v[0].get_value().unwrap(), "true".as_bytes().to_owned())
  }

  // #[test]
  // fn test_get_kv() {
  //   let client = Client::new("8.8.8.8", 8500);
  //   let my_keys = client.kv_get("my-key").unwrap();
  //   for k in my_keys {
  //     println!("k: {:?}", k);
  //   }
  // }
}


